lib/pbench/test/unit/agent/test_tool_group.py

```diff
@@ -268,3 +270,52 @@ def mock_read_text(path):
                 assert tg.get_label(host) == labels[host]
             assert tg.get_tools("bad-host") == {}
             assert tg.get_label("bad-host") == ""
+
+
+def test_gen_tool_groups():
+    """Verify the tool group directory generator."""
+
+    class MockPath:
+        """A simple ToolGroup mock to generate host names from different
+        tool group directories.
+        """
+
+        _groups = set(("heavy", "light", "medium"))
+
+        def __init__(self, name: str):
+            self._name = name
+
+        def glob(self, prefix: str):
+            if self._name == "no-tool-group-dirs":
+                # The special name which for a directory which "does not
+                # have tool groups".
+                return
+            for n in self._groups:
+                # Return the 3 tools groups that are found with the prefix
+                # attached.
+                yield Path(f"{prefix[:-2]}-{n}")
+
+    class MockToolGroup:
+        """A simple no-op ToolGroup class which just stores the constructor
+        arguments.
+        """
+
+        TOOL_GROUP_PREFIX = "tests-vX"
+
+        def __init__(self, name: str, pbench_run: str):
+            self.name = name
+            self.pbench_run = pbench_run
+
+    with mock.patch("pbench.agent.tool_group.Path", MockPath), mock.patch(
+        "pbench.agent.tool_group.ToolGroup", MockToolGroup
+    ):
+        # To verify the case where there are no tool groups created, we make a
+        # special Path object which will expose no tool groups.
+        tgs = list(gen_tool_groups("no-tool-group-dirs"))
+        assert len(tgs) == 0
+
+        # Using the special Path object which does have tool groups, we verify
+        # that the expected generated list of Tool Group objects matches the
+        # list of names we expect.
+        names = set([tg.name for tg in gen_tool_groups("have-tool-group-dirs")])
```

Comment:
`"have-tool-group-dirs"` is not a "special Path object".

It would be better if there were a nice symbolic definition for `"no-tool-group-dirs"`, e.g., `MockPath.NOSUBDIRS" -- then we would be able to get by without the comments.

Reply:
Looks like a difference between upper case and lower case references.  Is that really a change that is necessary?

Reply:
The difference is that `"no-tool-group-dirs"` is a literal string which has no scope or clue to its usage, while `MockPath.NOSUBDIRS` is an encapsulated value with a specific scope and reference to context.

That is, because `NOSUBDIRS` is qualified by a path (i.e., a class), the reader can refer to the class definition for context and potentially documentation.  No such (explicit) indicator is available for the literal string.  Also, because `NOSUBDIRS` encapsulates the value, we don't need to concern ourselves with what the actual "special Path object" _is_, and we can trivially change it if needed because it is specified in only one place and merely referenced in all the others.

The fact that this code snippet requires a block comment to explain what the string is and does (and the fact that the comment _here_ constrains or explains implementation **elsewhere**) suggests that we should probably be doing something different.  Using a symbolic constant instead of a literal value should address that and alleviate the need for the explanation of the _mechanism_ (if we still need to explain the intention, that's another story), and choosing a good name for the symbol might relieve the need for the rest of the comment (i.e., "code as documentation").

Reply:
I'll make a change, but I just don't see how this helps.  The symbols are used only within the scope of the test function so that it results in just upper case letters and underscores vs lowercase letters and dashes.  The comments are still useful, so this ends up just being more code.